src/header/common/strict_transport_security.rs

```diff
@@ -0,0 +1,304 @@
+            include_subdomains: false
+        }
+    }
+}
+
+enum StsDirective {
+    IncludeSubdomains(bool),
+    MaxAge(u64)
+}
+
+struct StsParser<'a> {
+    pos: usize,
+    input: &'a Vec<u8>
+}
+
+impl<'a> StsParser<'a> {
+    fn peek(&self) -> char {
+        *self.input[self.pos..].iter().next().unwrap() as char
+    }
+
+    fn eof(&self) -> bool {
+        self.pos >= self.input.len()
+    }
+
+    fn pop(&mut self) -> char {
+        let c = *self.input[self.pos..].iter().next().unwrap() as char;
+        self.pos += 1;
+        c
+    }
+
+    fn pop_while<F>(&mut self, test: F) -> &str where F: Fn(char) -> bool {
+        let start = self.pos;
+        while !self.eof() && test(self.peek()) {
+            self.pos += 1;
+        }
+
+        from_utf8(&self.input[start..self.pos]).unwrap()
+    }
+
+    fn pop_whitespace(&mut self) {
+        self.pop_while(|c: char| c.is_whitespace());
+    }
+
+    fn pop_to_sep(&mut self) {
+        { self.pop_while(|c| c != ';'); }
+
+        if !self.eof() {
+            self.pop();
+        }
+    }
+
+    fn parse_tokens(&mut self) -> &str {
+        { self.pop_whitespace(); }
+
+        let result = self.pop_while(|c| match c {
+            '(' | ')' | '<' | '>' | '@' | ',' | ';' | ':' | '\\' |
+            '"' | '/' | '[' | ']' | '?' | '=' | '{' | '}' | ' ' | '\t' => false,
+            _ => true
+        });
+
+        result
+    }
+
+    fn parse_directive_name(&mut self) -> &str {
+        self.parse_tokens()
+    }
+
+    fn parse_value(&mut self) -> &str {
+        self.parse_tokens()
+    }
+
+    fn parse_quoted_string(&mut self) -> &str {
+        self.pop();
+        let start = self.pos;
+        let mut end = self.pos;
+        let mut is_escaped = false;
+        loop {
+            if self.eof() {
+                break
+            } else if is_escaped {
+                is_escaped = false;
+                self.pos += 1;
+            } else if self.peek() == '"' {
+                end = self.pos;
+                self.pop();
+                break;
+            } else {
+                is_escaped = self.peek() == '\\';
+                self.pos += 1;
+            }
+        }
+
+        from_utf8(&self.input[start..end]).unwrap()
+    }
+
+    fn parse_directive_value(&mut self) -> Option<&str> {
+        self.pop_whitespace();
+        if self.eof() { return None }
+
+        match self.peek() {
+            '=' => {
+                self.pop();
+                { self.pop_whitespace(); }
+
+                if self.eof() { return None }
+
+                let result = match self.peek() {
+                    '"' => Some(self.parse_quoted_string()),
+                    _ => Some(self.parse_value())
+                };
+
+                result
+            }
+            _ => {
+                None
+            }
+        }
+    }
+
+    fn parse_directive(&mut self) -> Option<StsDirective> {
+        let directive_name = self.parse_directive_name().to_owned();
```

Comment:
If this is the allocation you are referring, I think you can convince the compiler by changing a number of the signatures to `fn foo(&mut self) -> &'a str` to unlink the return value's lifetime from the mutable borrow of self (lifetime elision) and instead link it to the structs buffer.